lazy-pages/src/common.rs

```diff
@@ -77,46 +76,104 @@ pub(crate) struct LazyPagesContext {
+    ) -> Result<(&LazyPagesRuntimeContext, &LazyPagesExecutionContext), ContextError> {
+        Ok((self.runtime_context()?, self.execution_context()?))
+    }
+
+    pub fn contexts_mut(
+        &mut self,
+    ) -> Result<(&mut LazyPagesRuntimeContext, &mut LazyPagesExecutionContext), ContextError> {
+        let rt_ctx = self
+            .runtime_context
+            .as_mut()
+            .ok_or(ContextError::RuntimeContextIsNotSet)?;
+        let exec_ctx = self
+            .execution_context
+            .as_mut()
+            .ok_or(ContextError::ExecutionContextIsNotSet)?;
+        Ok((rt_ctx, exec_ctx))
+    }
+
+    pub fn runtime_context(&self) -> Result<&LazyPagesRuntimeContext, ContextError> {
+        self.runtime_context
+            .as_ref()
+            .ok_or(ContextError::RuntimeContextIsNotSet)
+    }
+
     pub fn runtime_context_mut(&mut self) -> Result<&mut LazyPagesRuntimeContext, ContextError> {
         self.runtime_context
             .as_mut()
             .ok_or(ContextError::RuntimeContextIsNotSet)
     }
+
     pub fn execution_context(&self) -> Result<&LazyPagesExecutionContext, ContextError> {
         self.execution_context
             .as_ref()
             .ok_or(ContextError::ExecutionContextIsNotSet)
     }
-    pub fn execution_context_mut(
-        &mut self,
-    ) -> Result<&mut LazyPagesExecutionContext, ContextError> {
-        self.execution_context
-            .as_mut()
-            .ok_or(ContextError::ExecutionContextIsNotSet)
-    }
+
     pub fn set_runtime_context(&mut self, ctx: LazyPagesRuntimeContext) {
         self.runtime_context = Some(ctx);
     }
+
     pub fn set_execution_context(&mut self, ctx: LazyPagesExecutionContext) {
         self.execution_context = Some(ctx);
     }
 }
 
 pub(crate) type Weights = [u64; WeightNo::Amount as usize];
-pub(crate) type PageSizes = [NonZeroU32; PageSizeNo::Amount as usize];
 pub(crate) type GlobalNames = Vec<LimitedStr<'static>>;
 
+#[derive(Debug, Clone, Copy)]
+pub struct PageSizes(pub(crate) [NonZeroU32; PageSizeNo::Amount as usize]);
+
+impl SizeManager for PageSizes {
+    fn size_non_zero<P: PageDynSize>(&self) -> NonZeroU32 {
+        self.0[P::SIZE_NO]
+    }
+}
+
 #[derive(Debug)]
 pub(crate) struct LazyPagesRuntimeContext {
     pub page_sizes: PageSizes,
     pub global_names: GlobalNames,
     pub pages_storage_prefix: Vec<u8>,
+    pub program_storage: Box<dyn LazyPagesStorage>,
+}
+
+impl LazyPagesRuntimeContext {
+    pub fn page_has_data_in_storage(&self, prefix: &PagePrefix, page: GearPage) -> bool {
+        self.program_storage.page_exists(prefix, page)
+    }
+
+    pub fn load_page_data_from_storage(
+        &mut self,
+        prefix: &PagePrefix,
+        page: GearPage,
+        buffer: &mut [u8],
+    ) -> Result<bool, Error> {
+        self.program_storage
+            .load_page(&self.page_sizes, prefix, page, buffer)
+            .map_err(Error::PagesStorage)
+    }
+}
+
+pub trait LazyPagesStorage: fmt::Debug {
```

Comment:
why this trait is not in lazy-pages/common?

Reply:
Where should it be?

Reply:
you should change arguments `prefix: &PagePrefix, page: GearPage` to `page_key: &PageKey`. Where `PageKey` is associative type for trait LazyPagesStorage. So, you can remove that `parse_key` from gtest.

`PagePrefix` is just utils-struct for lazy-pages, it's strange to use it outside.

Reply:
Ok, as I see it's not comfortable here to use PagePrefix. But, will be better to use then one argument page_key: &[u8] (you can parse it to (ProgramId, GearPage) using codec in `gtest`. I also cannot see the reason to add there PAGE_STORAGE_PREFIX and memory_infix, they looks useless )